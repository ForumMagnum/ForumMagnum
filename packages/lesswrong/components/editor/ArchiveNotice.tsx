import React, {useState} from 'react';
import { Components, registerComponent } from '../../lib/vulcan-lib';
import ArchiveIcon from "@material-ui/icons/Archive";
import { Link, useNavigate } from '../../lib/reactRouterWrapper';
import classNames from 'classnames';
import IconButton from '@material-ui/core/IconButton';
import CloseIcon from '@material-ui/icons/Close';
import {useUpdate} from '../../lib/crud/withUpdate';
import {useLocation} from '../../lib/routeUtil';
import {useCurrentUser} from '../common/withUser';
import {userGetProfileUrl} from '../../lib/collections/users/helpers';
import {useMessages} from '../common/withMessages';


const styles = (theme: ThemeType): JssStyles => ({
  lwBanner: {
    padding: 16,
    backgroundColor: theme.palette.background.warningTranslucent,
    display: "flex",
    justifyContent: "space-between",
    alignItems: "flex-start",
    marginBottom: "60px",
  },
  hidden: {
    display: "none"
  },
  title: {
    fontWeight: 670,
  },
  icon: {
    marginLeft: 8,
    marginRight: 16,
    height: 24,
    color: theme.palette.grey[500]
  },
  buttonRow: {
    display: 'flex',
    displayDirection: 'row',
    alignItems: 'center',
  },
  buttonShared: {
    ...theme.typography.commentStyle,
     height: '36px',
     paddingLeft: 16,
     paddingRight: 16,
     marginRight: 10,
     color: 'white',
     fontSize: '14px',
     border: 0,
     fontWeight: '500',
     cursor: 'pointer',
     letterSpacing: '0.6',
     borderRadius: '6px',
     transition: 'all 0.2s ease',
     boxShadow: '0px 4px 5.5px 0px rgba(0, 0, 0, 0.07)',
     '&:hover': {
       opacity: 0.8
     },
  },
  restoreButton: {
    background: theme.palette.buttons.alwaysPrimary,
  },
  closeIcon: { 
    color: theme.palette.grey[500],
    opacity: 0.5,
    padding: 2,
  },
  subtleText: {
    color: theme.palette.text.dim3,
  }
});

// Tells the user when they can next comment or post if they're rate limited, and a brief explanation
const ArchiveNotice = ({post, classes}: {
  post: PostsPage, 
  classes: ClassesType
}) => {
  const { ContentStyles, Loading } = Components
  const [hidden, setHidden] = useState(false)
  const {flash} = useMessages();

  const {mutate: updatePost, loading, error} = useUpdate({
    collectionName: "Posts",
    fragmentName: 'PostsEdit',
  });

  const handleRestore = async () => {
    await updatePost({
      selector: {_id: post._id},
      data: {deletedDraft: false}
    })
    flash({
      messageString: "Restored", 
      type: "success", 
    })
  }

  if (loading) return <Loading />
  if (error) return <p>Error: {error.message}</p>

  return (
  <ContentStyles contentType="comment" className={classNames(classes.lwBanner, {[classes.hidden]: hidden})}>
    
      <div>
        <p className={classes.title}>This draft is archived</p>
          <p>It may either have been archived by a coauthor, or auto-archived due to inactivity.</p>
          <p className={classes.subtleText}>See all your archived drafts: <Link to={"/drafts?includeArchived=true"}>{"lesswrong.com/drafts"}</Link> 
        </p>
        <div className={classes.buttonRow}>
          <button className={classNames(classes.buttonShared, classes.restoreButton)} type="button" onClick={handleRestore}>
            Restore
          </button>
        </div>
      </div> 
    <IconButton className={classes.closeIcon} onClick={() => setHidden(true)}>
      <CloseIcon />
    </IconButton>
  </ContentStyles>)
}

const ArchiveNoticeComponent = registerComponent('ArchiveNotice', ArchiveNotice, {styles});

declare global {
  interface ComponentTypes {
    ArchiveNotice: typeof ArchiveNoticeComponent
  }
}
