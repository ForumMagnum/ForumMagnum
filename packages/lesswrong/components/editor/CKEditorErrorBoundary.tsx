import React, { Component, createContext } from 'react';
import { Components, registerComponent } from '../../lib/vulcan-lib';
import type { CKPostEditorProps } from './CKPostEditor';

interface ErrorState {
  error?: unknown;
}

interface CKEditorErrorBoundaryProps {
  handleError: (error: unknown) => void;
}

export class CKEditorErrorBoundary extends Component<CKEditorErrorBoundaryProps, ErrorState> {
  constructor(props: CKEditorErrorBoundaryProps) {
    console.log('in CKEditorErrorBoundary');
    super(props);
    this.state = {};
  }

  static getDerivedStateFromError(error: unknown) {
    console.log({ error }, 'in error boundary getDerivedStateFromError');
    return { error };
  }

  componentDidCatch(error: unknown) {
    console.error({ error }, 'in error boundary componentDidCatch');
    this.props.handleError(error);
    this.setState({ error });
  }

  render() {
    // const { CKPostEditor } = Components;
    // const { error } = this.state;

    // if (error) {
    //   console.log({ error }, 'rendering in CKEditorErrorBoundary');
    //   // this.props.handleError(error);
    //   return <>Error!</>;
    // }
    return this.props.children;
    // return (
    //   // <CKEditorErrorContext.Provider value={{ error }}>
    //     {this.props.children}
    //     {/* <CKPostEditor {...this.props} renderingError={error} /> */}
    //   // </CKEditorErrorContext.Provider>
    // );
  }
}

const CKEditorErrorBoundaryComponent = registerComponent('CKEditorErrorBoundary', CKEditorErrorBoundary);

declare global {
  interface ComponentTypes {
    CKEditorErrorBoundary: typeof CKEditorErrorBoundaryComponent
  }
}

