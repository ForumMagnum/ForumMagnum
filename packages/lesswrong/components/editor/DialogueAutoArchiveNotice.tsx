import React, {useState} from 'react';
import { Components, registerComponent } from '../../lib/vulcan-lib';
import moment from 'moment';
import { isEAForum } from '../../lib/instanceSettings';
import AlarmIcon from '@material-ui/icons/Alarm';
import { isFriendlyUI } from '../../themes/forumTheme';
import { Link, useNavigate } from '../../lib/reactRouterWrapper';
import classNames from 'classnames';
import IconButton from '@material-ui/core/IconButton';
import CloseIcon from '@material-ui/icons/Close';
import {useUpdate} from '../../lib/crud/withUpdate';
import {useLocation} from '../../lib/routeUtil';
import {useCurrentUser} from '../common/withUser';
import {userGetProfileUrl} from '../../lib/collections/users/helpers';
import {useMessages} from '../common/withMessages';


const styles = (theme: ThemeType): JssStyles => ({
  lwBanner: {
    padding: 12,
    backgroundColor: theme.palette.background.warningTranslucent,
    display: "flex",
    justifyContent: "space-between",
    alignItems: "flex-start",
    marginBottom: "60px",
  },
  hidden: {
    display: "none"
  },
  title: {
    fontWeight: 670,
  },
  icon: {
    marginLeft: 8,
    marginRight: 16,
    height: 24,
    color: theme.palette.grey[500]
  },
  buttonRow: {
    display: 'flex',
    displayDirection: 'row',
    alignItems: 'center',
  },
  buttonShared: {
    ...theme.typography.commentStyle,
     height: '36px',
     paddingLeft: 16,
     paddingRight: 16,
     marginRight: 10,
     color: 'white',
     fontSize: '14px',
     border: 0,
     fontWeight: '500',
     cursor: 'pointer',
     letterSpacing: '0.6',
     borderRadius: '6px',
     transition: 'all 0.2s ease',
     boxShadow: '0px 4px 5.5px 0px rgba(0, 0, 0, 0.07)',
     '&:hover': {
       opacity: 0.8
     },
  },
  okayButton: {
    background: theme.palette.buttons.alwaysPrimary,
  },
  archiveButton: {
    background: "none",
    color: theme.palette.text.primary,
    border: theme.palette.border.grey300,
  },
  closeIcon: { 
    color: theme.palette.grey[500],
    opacity: 0.5,
    padding: 2,
  },
  subtleText: {
    color: theme.palette.text.dim3,
  }
});

// Tells the user when they can next comment or post if they're rate limited, and a brief explanation
const DialogueAutoArchiveNotice = ({dialogue, daysLeft, classes}: {
  dialogue: PostsPage, 
  daysLeft: number,
  classes: ClassesType
}) => {
  const { ContentStyles, Loading } = Components
  const [hidden, setHidden] = useState(false)
  const navigate = useNavigate();
  const location = useLocation();
  const {flash} = useMessages();
  const currentUser = useCurrentUser();

  const {mutate: updatePost, loading, error} = useUpdate({
    collectionName: "Posts",
    fragmentName: 'PostsEdit',
  });

  const handleArchive = async () => {
    await updatePost({
      selector: {_id: dialogue._id},
      data: {deletedDraft: true}
    })
    const postUrl = location.location
    const profileUrl = userGetProfileUrl(currentUser)
    navigate(profileUrl)
    flash({
      messageString: "Dialogue archived", 
      type: "success", 
      action: () => { // Undo action
          void updatePost({
            selector: {_id: dialogue._id},
            data: {deletedDraft: false}
          })
          navigate(postUrl)
          flash({messageString: "Restoring... It's okay to change your mind :)", type: "success"})
      }
    })
  }

  if (loading) return <Loading />
  if (error) return <p>Error: {error.message}</p>

  return (
  <ContentStyles contentType="comment" className={classNames(classes.lwBanner, {[classes.hidden]: hidden})}>
    <AlarmIcon className={classes.icon} />

    <div>
      <p className={classes.title}>This dialogue will auto-archive in {daysLeft} days, unless there are new edits</p>
        <p>To keep your drafts orderly, we archive dialogues without activity. To cancel archiving, just make any edit below.</p>
        <p className={classes.subtleText}>See all your drafts, including archived: <Link to={"/drafts?includeArchived=true"}>{"lesswrong.com/drafts"}</Link> 
      </p>
      <div className={classes.buttonRow}>
        <button className={classNames(classes.buttonShared, classes.okayButton)} type="button" onClick={() => setHidden(true)}>
          Okay
        </button>
        <button className={classNames(classes.buttonShared, classes.archiveButton)} type="button" onClick={handleArchive}>
          Archive now
        </button>
      </div>
    </div>  
    
    <IconButton className={classes.closeIcon} onClick={() => setHidden(true)}>
      <CloseIcon />
    </IconButton>
  </ContentStyles>)
}

const DialogueAutoArchiveNoticeComponent = registerComponent('DialogueAutoArchiveNotice', DialogueAutoArchiveNotice, {styles});

declare global {
  interface ComponentTypes {
    DialogueAutoArchiveNotice: typeof DialogueAutoArchiveNoticeComponent
  }
}
