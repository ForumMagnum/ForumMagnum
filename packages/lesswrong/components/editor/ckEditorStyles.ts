import { defineStyles } from "../hooks/useStyles"

// Shared between post- and comment editors
export const ckEditorPluginStyles = defineStyles("CKEditor", (theme: ThemeType) => ({
  ckWrapper: {
    // Theme values exported for use in CkEditor-specific stylesheets
    "--palette-panelBackground-default": theme.palette.panelBackground.default,
    "--palette-fonts-sansSerifStack": theme.palette.fonts.sansSerifStack,
    "--palette-grey-200": theme.palette.grey[200],
    "--palette-grey-300": theme.palette.grey[300],
    "--palette-grey-600": theme.palette.grey[600],
    "--palette-grey-1000": theme.palette.grey[1000],
    "--palette-error-main": theme.palette.error.main,
    "--borderRadius-defaul": theme.borderRadius.default,
  },
  sidebar: {
    position: 'absolute',
    right: -350,
    width: 300,
    [theme.breakpoints.down('md')]: {
      position: 'absolute',
      right: -100,
      width: 50
    },
    [theme.breakpoints.down('sm')]: {
      right: 0
    }
  },
  addMessageButton: {
    marginBottom: 30,
  },
  hidden: {
    display: "none",
  },
}))
