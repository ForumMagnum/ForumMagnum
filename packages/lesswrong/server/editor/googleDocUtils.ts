import axios from 'axios';
import cheerio from 'cheerio';
import Jimp from 'jimp';
import { convertImagesInHTML, uploadBufferToCloudinary } from '../scripts/convertImagesToCloudinary';
import { extractTableOfContents } from '@/lib/tableOfContents';
import { dataToCkEditor } from './conversionUtils';
import { parseDocumentFromString } from '@/lib/domParser';

/**
 * Convert the footnotes we get in google doc html to a format ckeditor can understand. This is mirroring the logic
 * in the footnotes plugin (see e.g. ckEditor/src/ckeditor5-footnote/src/footnoteEditing/googleDocsFootnotesNormalizer.js)
 */
function googleDocConvertFootnotes(html: string): string {
  const $ = cheerio.load(html);

  const footnotePattern = /^ftnt(\d+)$/; // The actual footnotes at the bottom of the doc
  const referencePattern = /^ftnt_ref(\d+)$/; // The links to the footnotes in the main text

  const footnotes: Record<string, { item: cheerio.Cheerio; anchor: cheerio.Cheerio, id: string }> = {};
  $('a[id]').each((_, element) => {
    if (!('attribs' in element)) return

    const match = element.attribs.id.match(footnotePattern);
    if (match) {
      const index = match[1];
      // Find the closest parent div of the footnote anchor
      const footnoteDiv = $(element).closest('div');
      footnotes[index] = {
        item: footnoteDiv,
        anchor: $(element),
        id: Math.random().toString(36).slice(2),
      };
    }
  });

  if (Object.keys(footnotes).length === 0) {
    return $.html();
  }

  const references: Record<string, { item: cheerio.Cheerio; id: string }> = {};
  $('a[id]').each((_, element) => {
    if (!('attribs' in element)) return

    const match = element.attribs.id.match(referencePattern);
    if (match) {
      const index = match[1];
      if (footnotes.hasOwnProperty(index)) {
        references[index] = {
          item: $(element),
          id: footnotes[index].id,
        };
      }
    }
  });

  // Normalize the references by adding attributes and replacing the original <sup> tag
  Object.entries(references).forEach(([index, { item, id }]) => {
    item.attr('data-footnote-reference', '');
    item.attr('data-footnote-index', index);
    item.attr('data-footnote-id', id);
    item.attr('role', 'doc-noteref');
    item.parents('sup').replaceWith(item); // Replace the original <sup> tag with `item`
    item.wrap(`<span class="footnote-reference" id="fnref${id}"></span>`);
    item.text(`[${index}]`);
  });

  // Create the footnotes section
  $('body').append('<ol class="footnote-section footnotes" data-footnote-section="" role="doc-endnotes"></ol>');

  // Normalize the footnotes and put them in the newly created section
  Object.entries(footnotes).forEach(([index, { item, anchor, id }]) => {
    // Remove e.g. "[1]&nbsp;" from the footnote
    const anchorHtml = anchor.html();
    if (anchorHtml && anchorHtml.startsWith('&nbsp;')) {
      anchor.html(anchorHtml.replace(/^&nbsp;/, ''));
    } else {
      anchor.remove();
    }

    const footnoteContent = item.clone().addClass('footnote-content');

    // Replace bullets in footnotes with regular <p> elements and move them up a level
    const listParent = footnoteContent.find('li').parent('ul, ol');
    listParent.children('li').each((_, liElement) => {
      const liContent = $(liElement).html();
      $(`<p>${liContent}</p>`).insertBefore(listParent);
    });
    listParent.remove();

    const newFootnoteBackLink = $('<span class="footnote-back-link" data-footnote-back-link="" data-footnote-id="' + id + '"><sup><strong><a href="#fnref' + id + '">^</a></strong></sup></span>');

    const newFootnoteContent = $('<div class="footnote-content" data-footnote-content="" data-footnote-id="' + id + '" data-footnote-index="' + index + '"></div>');
    newFootnoteContent.append(footnoteContent.contents());

    const newFootnoteItem = $('<li class="footnote-item" data-footnote-item="" data-footnote-id="' + id + '" data-footnote-index="' + index + '" role="doc-endnote" id="fn' + id + '"></li>');
    newFootnoteItem.append(newFootnoteBackLink);
    newFootnoteItem.append(newFootnoteContent);

    $('.footnote-section').append(newFootnoteItem);
    item.remove()
  });

  // The changes so far leave over a stub like so:
  // <hr />
  //   <div>
  //     <p></p>
  //   </div>
  //   ...
  // <ol class=\"footnotes\" role=\"doc-endnotes\">
  //
  // Remove everything from the <hr /> to the footnotes section
  const footnotesSection = $('.footnote-section');
  const hrBeforeFootnotes = footnotesSection.prevAll('hr').last();
  hrBeforeFootnotes.remove();

  return $.html();
}

/**
 * Remote the google redirect from links that come from google docs
 *
 * https://www.google.com/url?q=https://en.wikipedia.org/wiki/Main_Page becomes https://en.wikipedia.org/wiki/Main_Page
 */
function googleDocRemoveRedirects(html: string): string {
  const $ = cheerio.load(html);

  // Regex match examples:
  // https://www.google.com/url?q=https://en.wikipedia.org/wiki/Main_Page&sa=D&source=editors&ust=1667922372715536&usg=AOvVaw2NyT5CZhfsrRY_zzMs2UUJ
  //                              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ <- first match group matches this, stopping at the first &
  // https://www.google.com/url?q=https://en.wikipedia.org/wiki/Main_Page
  //                              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ <- if there are no more params (no &), match up to the end of the string
  const hrefPattern = /^https:\/\/www\.google\.com\/url\?q=(\S+?)(&|$)/;

  $('a[href]').each((_, element) => {
    const href = $(element).attr('href');
    if (!href) return

    const match = hrefPattern.exec(href);
    if (match && match[1]) {
      $(element).attr('href', decodeURIComponent(match[1]));
    }
  });

  return $.html();
}

/**
 * Converts Google Docs formatting to ckeditor formatting. Currently handles:
 * - Italics
 * - Bold
 */
function googleDocTextFormatting(html: string): string {
  const $ = cheerio.load(html);

  $('span').each((_, element) => {
    const span = $(element);
    const fontStyle = span.css('font-style');
    const fontWeight = span.css('font-weight');

    if (fontStyle === 'italic' && fontWeight === '700') {
      span.wrap('<i><strong></strong></i>');
    } else if (fontStyle === 'italic') {
      span.wrap('<i></i>');
    } else if (fontWeight === '700') {
      span.wrap('<strong></strong>');
    }
  });

  return $.html();
}

/**
 * Convert the CSS based "cropping" in the imported html into actual cropping
 */
async function googleDocCropImages(html: string): Promise<string> {
  // Example of CSS-based cropping:
  // <p>
  //   <span style="overflow: hidden; display: inline-block; width: 396.00px; height: 322.40px;">
  //     <img src="https://example.com/image.jpg"
  //          style="width: 602.00px; height: 427.97px; margin-left: -110.00px; margin-top: -48.60px;">
  //   </span>
  // <p>

  const $ = cheerio.load(html);

  const cropPromises = $('p > span:has(> img)').map(async (_, span) => {
    const $span = $(span);
    const $img = $span.find('img');

    if ($img.length === 0) return;

    const spanStyle = $span.attr('style');
    const imgStyle = $img.attr('style');
    const src = $img.attr('src');

    if (!spanStyle || !imgStyle || !src) return;

    const spanWidth = parseFloat(spanStyle.match(/width:\s*([\d.]+)px/)?.[1] || '0');
    const spanHeight = parseFloat(spanStyle.match(/height:\s*([\d.]+)px/)?.[1] || '0');
    const imgWidth = parseFloat(imgStyle.match(/width:\s*([\d.]+)px/)?.[1] || '0');
    const imgHeight = parseFloat(imgStyle.match(/height:\s*([\d.]+)px/)?.[1] || '0');
    const marginLeft = parseFloat(imgStyle.match(/margin-left:\s*([-\d.]+)px/)?.[1] || '0');
    const marginTop = parseFloat(imgStyle.match(/margin-top:\s*([-\d.]+)px/)?.[1] || '0');

    if (imgWidth === 0 || imgHeight === 0) {
      return
    }

    const leftRelative = Math.max(-marginLeft, 0) / imgWidth;
    const topRelative = Math.max(-marginTop, 0) / imgHeight;
    const widthRelative = Math.round(spanWidth) / imgWidth;
    const heightRelative = Math.round(spanHeight) / imgHeight;

    if (leftRelative === 0 && topRelative === 0 && widthRelative === 1 && heightRelative === 1) {
      return
    }

    try {
      const response = await axios.get(src, { responseType: 'arraybuffer' });
      const buffer = Buffer.from(response.data, 'binary');

      const image = await Jimp.read(buffer);
      const originalWidth = image.bitmap.width;
      const originalHeight = image.bitmap.height;

      if (!originalWidth || !originalHeight) {
        throw new Error(`width or height not defined for image`)
      }

      const leftPixels = Math.round(leftRelative * originalWidth)
      const topPixels = Math.round(topRelative * originalHeight)
      const widthPixels = Math.min(Math.round(widthRelative * originalWidth), originalWidth - leftPixels)
      const heightPixels = Math.min(Math.round(heightRelative * originalHeight), originalHeight - topPixels)

      const croppedImage = await image.crop(leftPixels, topPixels, widthPixels, heightPixels);
      const croppedBuffer = await croppedImage.getBufferAsync(image.getMIME());

      const url = await uploadBufferToCloudinary(croppedBuffer)

      if (!url) {
        throw new Error(`Failed to upload cropped image to cloudinary`)
      }

      $img.attr('src', url);
      $img.attr('style', `width: ${widthPixels}px; height: ${heightPixels}px;`);
      $span.replaceWith($img);
    } catch (error) {
      // eslint-disable-next-line no-console
      console.error(`Error cropping image, falling back to uncropped version. src: ${src}`, error);
    }
  }).get();

  await Promise.all(cropPromises);

  return $.html();
}

/**
 * Removes comments from the raw html exported from Google Docs.
 */
function googleDocStripComments(html: string): string {
  const $ = cheerio.load(html);

  // Remove any <sup> tags that contain a child with a #cmnt id
  $('sup').each((_, element) => {
    const sup = $(element);
    if (sup.find('a[id^="cmnt_ref"]').length > 0) {
      sup.remove();
    }
  });

  // Remove the whole box at the bottom containing the comments
  $('div').each((_, element) => {
    const div = $(element);
    if (div.find('a[id^="cmnt"]').length > 0) {
      div.remove();
    }
  });

  return $.html();
}

/**
 * Converts internal links in Google Docs HTML to a format with `data-internal-id` attributes on block level elements.
 * This is used to maintain internal document links when importing into ckeditor, which doesn't use `id` attributes
 * on elements for internal linking.
 */
async function googleDocInternalLinks(html: string): Promise<string> {
  const $ = cheerio.load(html);

  // Define block level elements that are considered as blocks in ckeditor
  const blockLevelElements = ['p', 'div', 'h1', 'h2', 'h3', 'h4', 'h5', 'h6', 'ol', 'ul', 'li', 'blockquote', 'pre', 'hr', 'table'];

  // Map id to data-internal-id
  $('[id]').each((_, element) => {
    // Remove the id attribute and store its value
    const idValue = $(element).attr('id');
    $(element).removeAttr('id');

    // Find the nearest parent that is a block level element
    const blockParent = $(element).closest(blockLevelElements.join(','));

    if (blockParent.length && idValue) {
      blockParent.attr('data-internal-id', idValue)
    }
  });

  const tocHtml = (await extractTableOfContents(parseDocumentFromString($.html())))?.html;
  if (tocHtml) {
    const idMap: Record<string, string> = {};
    const $toc = cheerio.load(tocHtml);
    $toc('[id][data-internal-id]').each((_, element) => {
      const readableId = $toc(element).attr('id');
      const internalId = $toc(element).attr('data-internal-id');
      if (readableId && internalId) {
        idMap[internalId] = readableId;
      }
    });

    // Update all hrefs to point to the readable ids
    $('a[href]').each((_, element) => {
      const href = $(element).attr('href');
      if (href && href.startsWith('#')) {
        const internalId = href.slice(1);
        if (idMap[internalId]) {
          $(element).attr('href', `#${idMap[internalId]}`);
        }
      }
    });

    // Update the data-internal-ids to be equal to the readable id, this
    // means that even if they change the heading the links will still work
    $('[data-internal-id]').each((_, element) => {
      const internalId = $(element).attr('data-internal-id');
      if (internalId && idMap[internalId]) {
        $(element).attr('data-internal-id', idMap[internalId]);
      }
    });
  }

  return $.html();
}

/**
 * Handle footnotes and internal links. These are bundled together because they must
 * be done in the right order
 */
function googleDocConvertLinks(html: string) {
  const withNormalizedFootnotes = googleDocConvertFootnotes(html);
  const withInternalLinks = googleDocInternalLinks(withNormalizedFootnotes);
  return withInternalLinks
}

/**
 * In google docs nested bullets are handled through styling (each indentation level is actually
 * a separate list, with a margin-left creating the effect of nesting). In ckeditor, nested bullets are
 * actually <ul>s nested within each other. Convert this styling based nesting to genuine nesting
 */
function googleDocConvertNestedBullets(html: string): string {
  const $ = cheerio.load(html);

  // Each nesting level (<ul> or <ol> group) has a class like lst-kix_gwukp0509sil-0, or lst-kix_gwukp0509sil-1
  // The number at the end indicates the level of indentation, and the group a nesting level corresponds to
  // can be inferred from the fact that it is part of a continuous block of <ul>/<ol>s with nothing in between
  const listGroups: Record<string, {element: cheerio.Cheerio, index: number}[]> = {};
  let currentGroupId = 0;
  let lastListElement: cheerio.Element | null = null;

  $('ul[class*="lst-"], ol[class*="lst-"]').each((_, element) => {
    // If the current list element is not immediately after the last one, it's a new group
    if (!lastListElement || (element.prev && !$(element.prev).is(lastListElement))) {
      currentGroupId++;
    }
    lastListElement = element;

    const classNames = $(element).attr('class')?.split(/\s+/);
    const listClass = classNames?.find(name => name.startsWith('lst-'));
    if (!listClass) return;

    const match = listClass.match(/lst-([a-z_0-9]+)-(\d+)/);
    if (!match) return;

    const [ , , index] = match;
    if (!listGroups[currentGroupId]) {
      listGroups[currentGroupId] = [];
    }
    listGroups[currentGroupId].push({ element: $(element), index: parseInt(index) });
  });

  // Adjust the indices to account for contraints in ckeditor, and convert to genuine nesting
  for (const group of Object.values(listGroups)) {
    // In ckeditor, lists aren't allowed to start indented
    group[0].index = 0;

    // Indices can only increase by 1 from element to element
    for (let i = 1; i < group.length; i++) {
      if (group[i].index > group[i-1].index + 1) {
        group[i].index = group[i-1].index + 1;
      }
    }

    // Convert to genuine nesting
    group.forEach((item, i, arr) => {
      if (i > 0) {
        const prevItem = arr[i - 1];
        if (item.index === prevItem.index + 1) {
          prevItem.element.children('li:last-child').append(item.element);
        } else if (item.index <= prevItem.index) {
          // Find the ancestor list that matches the current index
          let ancestor = prevItem.element;
          for (let j = 0; j < prevItem.index - item.index; j++) {
            ancestor = ancestor.parent().closest('ul, ol');
          }
          ancestor.after(item.element);
        }
      }
    });
  };

  return $.html();
}

/**
 * To fix double spacing, remove all empty <p> tags that are immediate children of <body> from the HTML
 */
function removeEmptyBodyParagraphs(html: string): string {
  const $ = cheerio.load(html);

  $('body > p').each((_, element) => {
    const p = $(element);
    // Allow otherwise empty paragraphs containing images
    if (p.text().trim() === '' && p.find('img').length === 0) {
      p.remove();
    }
  });

  return $.html();
}

/**
 * We need to convert a few things in the raw html exported from google to make it work with ckeditor, this is
 * largely mirroring conversions we do on paste in the ckeditor code:
 * - Convert footnotes to our format
 * - Remove google redirects from all urls
 * - Reupload images to cloudinary (we actually don't do this on paste, but it's easier to do so here and prevents images breaking due to rate limits)
 */
export async function convertImportedGoogleDoc({
  html,
  postId,
}: {
  html: string;
  postId: string;
}) {
  const converters: (((html: string) => Promise<string>) | ((html: string) => string))[] = [
    async (html: string) => {
      const { html: rehostedHtml } = await convertImagesInHTML(html, postId, (url) =>
        url.includes("googleusercontent")
      );
      return rehostedHtml;
    },
    googleDocStripComments,
    googleDocCropImages,
    googleDocTextFormatting,
    googleDocConvertLinks,
    googleDocRemoveRedirects,
    googleDocConvertNestedBullets, // Must come before removeEmptyBodyParagraphs because paragraph breaks are used to determine when to break up a nested list of bullets
    removeEmptyBodyParagraphs,
    async (html: string) => await dataToCkEditor(html, "html"),
  ];

  let result: string = html;
  // Apply each converter in sequence
  for (const converter of converters) {
    result = await converter(result);
  }

  return result;
}
