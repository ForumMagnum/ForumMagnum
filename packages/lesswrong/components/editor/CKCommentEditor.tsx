import React from 'react'
import { registerComponent } from '../../lib/vulcan-lib/components';
import CKEditor from '../../lib/vendor/ckeditor5-react/ckeditor';
import { getCkEditor, ckEditorBundleVersion } from '../../lib/wrapCkEditor';
import { generateTokenRequest } from '../../lib/ckEditorUtils';
import { ckEditorUploadUrlSetting, ckEditorWebsocketUrlSetting } from '../../lib/publicSettings'
import { ckEditorUploadUrlOverrideSetting, ckEditorWebsocketUrlOverrideSetting } from '../../lib/instanceSettings';
import { defaultEditorPlaceholder } from '../../lib/editor/make_editable';
import { mentionPluginConfiguration } from "../../lib/editor/mentionsConfig";

// Uncomment the import and the line below to activate the debugger
// import CKEditorInspector from '@ckeditor/ckeditor5-inspector';

const CKCommentEditor = ({
  data,
  collectionName,
  fieldName,
  onSave,
  onChange,
  onFocus,
  onInit,
  placeholder,
}: {
  data?: any,
  collectionName: CollectionNameString,
  fieldName: string,
  onSave?: any,
  onChange?: any,
  onFocus?: (event: AnyBecauseTodo, editor: AnyBecauseTodo) => void,
  onInit?: any,
  placeholder?: string,
}) => {
  const webSocketUrl = ckEditorWebsocketUrlOverrideSetting.get() || ckEditorWebsocketUrlSetting.get();
  const ckEditorCloudConfigured = !!webSocketUrl;
  const { CommentEditor } = getCkEditor();

  return <div>
    <CKEditor
      editor={CommentEditor}
      onReady={(editor: any) => {
        // Uncomment the line below and the import above to activate the debugger
        // CKEditorInspector.attach(editor)
        if (onInit) onInit(editor)
        return editor
      }}
      onChange={onChange}
      onFocus={onFocus}
      config={{
        cloudServices: ckEditorCloudConfigured ? {
          // A tokenUrl token is needed here in order for image upload to work.
          // (It's accessible via drag-and-drop onto the comment box, and is
          // stored on CkEditor's CDN.)
          //
          // The collaborative editor is not activated because no `websocketUrl`
          // or `documentId` is provided.
          tokenUrl: generateTokenRequest(collectionName, fieldName),
          uploadUrl: ckEditorUploadUrlOverrideSetting.get() || ckEditorUploadUrlSetting.get(),
          bundleVersion: ckEditorBundleVersion,
        } : undefined,
        autosave: {
          save (editor: any) {
            return onSave && onSave( editor.getData() )
          }
        },
        initialData: data || "",
        placeholder: placeholder ?? defaultEditorPlaceholder,
        mention: mentionPluginConfiguration
      } as any}
      data={data}
    />
  </div>
}

const CKCommentEditorComponent = registerComponent("CKCommentEditor", CKCommentEditor);
declare global {
  interface ComponentTypes {
    CKCommentEditor: typeof CKCommentEditorComponent
  }
}
