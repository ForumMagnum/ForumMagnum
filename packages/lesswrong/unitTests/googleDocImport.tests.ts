import { convertImportedGoogleDoc } from '@/server/editor/googleDocUtils';

describe("convertImportedGoogleDoc", () => {

  it("Regression: Handle nested bullets with different list ids", async () => {
    const htmlInput = `
      <html>
        <body>
          <p><span></span></p>
          <p><span>Paragraph</span></p>
          <ul class="lst-kix_gia5zvmdml9o-0 start">
            <li><span>Bullet 1</span></li>
            <li><span>Bullet 2</span></li>
          </ul>
          <h2 data-internal-id="Header"><span>Header</span></h2>
          <ul class="lst-kix_82zwahk8jahu-0 start">
            <li><span>Start of a second list</span></li>
          </ul>
          <ul class="lst-kix_gia5zvmdml9o-0">
            <li><span>Bullet copy-pasted from the first list</span></li>
          </ul>
          <ul class="lst-kix_gia5zvmdml9o-1 start">
            <li><span>Sub-bullet 1</span></li>
            <li><span>Sub-bullet 2</span></li>
          </ul>
        </body>
      </html>
    `;
    const htmlOutput = await convertImportedGoogleDoc({html: htmlInput.replace(/\s+</g, '<'), postId: 'dummy'});
    expect(htmlOutput).toMatchSnapshot();
  });
});
