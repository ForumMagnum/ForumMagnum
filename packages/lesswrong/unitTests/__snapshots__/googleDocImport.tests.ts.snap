// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`convertImportedGoogleDoc Regression: Handle nested bullets with different list ids 1`] = `
"<p><span>Paragraph</span></p><ul><li><span>Bullet 1</span></li><li><span>Bullet 2</span></li></ul><h2 data-internal-id="Header"><span>Header</span></h2><ul><li><span>Start of a second list</span></li></ul><ul><li><span>Bullet copy-pasted from the first list</span><ul><li><span>Sub-bullet 1</span></li><li><span>Sub-bullet 2</span></li></ul></li></ul>
    "
`;
